src/data/equiv/algebra.lean

```diff
@@ -325,42 +344,111 @@ def map_equiv (h : α ≃* β) : units α ≃* units β :=
-{ hom := { map_one := e.to_mul_equiv.symm.map_one,
-           map_mul := e.to_mul_equiv.symm.map_mul,
-           map_add := e.to_add_equiv.symm.map_add },
-  .. e.to_equiv.symm }
+/-- The identity map is a ring isomorphism. -/
+@[refl] protected def refl : α ≃+* α := { .. mul_equiv.refl α, .. add_equiv.refl α }
 
-protected def trans {α β γ : Type*} [ring α] [ring β] [ring γ]
-  (e₁ : α ≃r β) (e₂ : β ≃r γ) : α ≃r γ :=
-{ hom := is_ring_hom.comp _ _, .. e₁.to_equiv.trans e₂.to_equiv  }
+variables {α}
 
-instance symm.is_ring_hom {e : α ≃r β} : is_ring_hom e.to_equiv.symm := hom e.symm
+/-- The inverse of a ring isomorphism is a ring isomorphis. -/
+@[symm] protected def symm (e : α ≃+* β) : β ≃+* α :=
+{ .. e.to_mul_equiv.symm, .. e.to_add_equiv.symm }
 
-@[simp] lemma to_equiv_symm (e : α ≃r β) : e.symm.to_equiv = e.to_equiv.symm := rfl
+/-- Transitivity of `ring_equiv`. -/
+@[trans] protected def trans (e₁ : α ≃+* β) (e₂ : β ≃+* γ) : α ≃+* γ :=
+{ .. (e₁.to_mul_equiv.trans e₂.to_mul_equiv), .. (e₁.to_add_equiv.trans e₂.to_add_equiv) }
 
-@[simp] lemma to_equiv_symm_apply (e : α ≃r β) (x : β) :
+@[simp] lemma to_equiv_symm (e : α ≃+* β) : e.symm.to_equiv = e.to_equiv.symm := rfl
+
+@[simp] lemma to_equiv_symm_apply (e : α ≃+* β) (x : β) :
   e.symm.to_equiv x = e.to_equiv.symm x := rfl
 
+end basic
+
+section
+
+variables [semiring α] [semiring β] (f : α ≃+* β) (x y : α)
+
+/-- A ring isomorphism preserves multiplication. -/
+lemma map_mul : f (x * y) = f x * f y := f.map_mul' x y
+
+/-- A ring isomorphism sends one to one. -/
+lemma map_one : f 1 = 1 := (f : α ≃* β).map_one
+
+/-- A ring isomorphism preserves addition. -/
+lemma map_add : f (x + y) = f x + f y := f.map_add' x y
+
+/-- A ring isomorphism sends zero to zero. -/
+lemma map_zero : f 0 = 0 := (f : α ≃+ β).map_zero
+
+end
+
+section
+
+variables [ring α] [ring β] (f : α ≃+* β) (x y : α)
+
+lemma map_neg : f (-x) = -f x := (f : α ≃+ β).map_neg x
+
+lemma map_sub : f (x - y) = f x - f y := (f : α ≃+ β).map_sub x y
+
+lemma map_neg_one : f (-1) = -1 := f.map_one ▸ f.map_neg 1
+
+end
+
+section semiring_hom
+
+variables [semiring α] [semiring β]
+
+/-- Reinterpret a ring equivalence as a ring homomorphism. -/
+def to_ring_hom (e : α ≃+* β) : α →+* β :=
+{ .. e.to_mul_equiv.to_monoid_hom, .. e.to_add_equiv.to_add_monoid_hom }
+
+/-- Interpret an equivalence `f : α ≃ β` as a ring equivalence `α ≃+* β`. -/
+def of (e : α ≃ β) [is_semiring_hom e] : α ≃+* β :=
+{ .. e, .. monoid_hom.of e, .. add_monoid_hom.of e }
+
+instance (e : α ≃+* β) : is_semiring_hom e := e.to_ring_hom.is_semiring_hom
+
+end semiring_hom
+
+section ring_hom
+
+variables [ring α] [ring β]
+
+/-- Interpret an equivalence `f : α ≃ β` as a ring equivalence `α ≃+* β`. -/
+def of' (e : α ≃ β) [is_ring_hom e] : α ≃+* β :=
+{ .. e, .. monoid_hom.of e, .. add_monoid_hom.of e }
+
+instance (e : α ≃+* β) : is_ring_hom e := e.to_ring_hom.is_ring_hom
+
+instance is_ring_hom' (e : α ≃+* β) : is_ring_hom e.to_equiv.to_fun :=
```

Comment:
Wouldn't it be better to (also) have a (simp)-lemma that puts these uncoerced forms back into "coe-normal-form"?

Reply:
Removed this instance, fixed the only file that used this instance.